ENH: Improve error message for empty object array

bashtage opened the pull request:
Improve the error message shown when an object array is empty

closes #23572

- [x] closes #23572
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry


No what's new since too small a change and no test needed.

bashtage committed:
ENH: Improve error message for empty object array

Improve the error message shown when an object array is empty

closes #23572

pep8speaks commented:
Hello @bashtage! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/io/stata.py`](https://github.com/bashtage/pandas/blob/3eb95fd30d00ce69526a7a422a939d8a3b6772e4/pandas/io/stata.py) !

gfyoung commented on pandas/io/stata.py:
```diff
@@ -1868,7 +1868,12 @@ def _dtype_to_default_stata_fmt(dtype, column, dta_version=114,
+                             'exported. Object arrays containing only null '
+                             'values are prohibited. Other object types'
+                             'cannot be exported and must first be converted '
+                             'to one of the supported types.')
```
* By one of the supported types, do you just mean string?
* I presume a test for this error message is coming?

bashtage replied:
> By one of the supported types, do you just mean string?

It could be any supported type, int8-32, float, double, string or datetime. The user has to make a choice as to what best expresses the object.

bashtage committed:
TST: Add tests for all None

Test exception is hit when all values in an object column are None
Extend the test for strl conversion to ensure this case passes (as expected)

bashtage commented:
I also extended the test in my previous PR to ensure that the case with all `None` when specifically requested to write as a strl passes.  I hope this is OK, but could move to a new PR is needed.

jtkiley commented:
@bashtage Consider adding the offending column to the error like in #23646. The all `None` case isn't hard to handle, but most of the work for me is identifying the column(s) that are the problem in datasets with hundreds of columns.

Since it's throwing on a particular column, it may be that the first encountered column is the only one that can be named, but that's still helpful. An alternative would be providing a brief code example that will identify these all `None` columns (and perhaps an example of fixing it with `convert_strl`).

I could imagine relatively novice pandas users (i.e. Stata switchers) hitting this error, so I'd lean toward a little more hand-holding/direction. Happy to hear counterarguments, though.

bashtage commented:
@jtkiley I added this.

jtkiley commented:
@bashtage Thanks for this (looks great) and all you do on the Stata format. It's a big help for those of us working with both, and it helps smooth the work I do to train Stata users to use pandas.

jreback commented:
thanks @bashtage
